refactor: bitcoin-config.h includes cleanup

theuni opened the pull request:
As mentioned in https://github.com/bitcoin/bitcoin/pull/26924#issuecomment-1403449932 and https://github.com/bitcoin/bitcoin/pull/29263#issuecomment-1922334399, it is currently not safe to remove `bitcoin-config.h` includes from headers because some unrelated file might be depending on it.

See also #26972 for discussion.

Solve this by including the file directly everywhere it's required, regardless of whether or not it's already included by another header.

There should be no functional change here, but it will allow us to safely remove includes from headers in the future.

~I'm afraid it's a bit tedious to reproduce these commits, but it's reasonably straightforward:~
Edit: See note below

```bash
# All commands executed from the src/ subdir.

# Collect all tokens from bitcoin-config.h.in
# Isolate the tokens and remove blank lines
# Replace newlines with | and remove the last trailing one
# Collect all files which use these tokens
# Filter out subprojects (proper forwarding can be verified from Makefiles)
# Filter out .rc files
# Save to a text file
git grep -E -l `grep undef config/bitcoin-config.h.in | cut -d" " -f2 | grep -v '^$' | tr '\n' '|' | sed 's/|$//'` | grep -v -e "^leveldb/" -e "^secp256k1/" -e "^crc32c/" -e "^minisketch/" -e "^Makefile" -e "\.rc$" > files-with-config-include.txt

# Find all files from the above list which don't include bitcoin-config.h
git grep -L -E "config/bitcoin-config.h" -- `cat files-with-config-include.txt`

# Include them manually with the exception of some files in crypto:
# crypto/sha256_arm_shani.cpp crypto/sha256_avx2.cpp crypto/sha256_sse41.cpp crypto/sha256_x86_shani.cpp
# These are exceptions which don't use bitcoin-config.h, rather the Makefile.am adds these cppflags manually.

# Commit changes. This should match the first commit of this PR.

# Use the same search as above to find all files which DON'T use any config tokens
git grep -E -L `grep undef config/bitcoin-config.h.in | cut -d" " -f2 | grep -v '^$' | tr '\n' '|' | sed 's/|$//'` | grep -v -e "^leveldb/" -e "^secp256k1/" -e "^crc32c/" -e "^minisketch/" -e "^Makefile" -e "\.rc$" > files-without-config-include.txt

# Manually remove the includes and commit changes. This should match the second commit of this PR.
```

Edit: I'll keep this old description for posterity, but the manual approach has been replaced with a scripted diff from TheCharlatan

theuni commented:
Ping @maflcko

This (or something like it) is a prerequisite for #29404.

There doesn't seem to be any sane way to make this a scripted diff or a clang-tidy transform, so it would be great if someone could reproduce my results.

theuni commented:
Removed the include from `interfaces/wallet.h` which snuck in because of a comment:
```c++
//! function will be undefined in builds where ENABLE_WALLET is false.
```
That was the only example of a false-positive that I managed to find.

TheCharlatan commented:
I validated the patch by running:
\> `iwyu_tool -p . -j $(nproc) -- -Xiwyu --max_line_length=180 -I/usr/lib/llvm-14/lib/clang/14/include/ -Xiwyu --mapping_file=/home/drgrid/bitcoin/contrib/devtools/iwyu/bitcoin.core.imp > iwyu_output.txt`
\> `cat iwyu_output.txt | awk '/should remove these lines:/ {file=$1} /- #include <config\/bitcoin-config.h>/ {split($NF, a, "-"); line=a[1]; if (!seen[file, line]++) print file " " a[1]}' | while read -r file line; do sed -i -e "$((line-1)),$((line+2))d" "src/$file"; done;`
\> `cat iwyu_output.txt | awk '/should add these lines:/ {file=$1} /^#include "bitcoin-config.h"/ {if (!seen[file]++) print file ";" "#include <config/bitcoin-config.h>" " " substr($0, index($0, "//"))}' | while IFS=";" read -r file include; do sed -i "/\/\/ file COPYING or http:\/\/www.opensource.org\/licenses\/mit-license.php\./a \\\\n\#if defined(HAVE_CONFIG_H)\\n$include\\n\#endif" "src/$file" ; done`

The script misses a few of the cases, because some subset of the `bitcoin-config.h` symbols are left commented or undefined, meaning IWYU is not capable of picking up on them. It does add the commented required symbol names though, which should be part of this patch in my opinion. I'll try to come back with a scripted-diff solution for this.

Maybe a tidy plugin doing a full text search during the pre-processor hook could work, which would be kind of similar to the process you already posted here. So I guess we could add a little rust lint job for this instead?

maflcko reviewed (commented):
> Include them manually with the exception of some files in crypto:

unrelated question: Is there a reason they need to use the flag at all? The files are never compiled and linked when the flag isn't set, so making them appear "empty" when the flag isn't set is not needed?

theuni commented:
> > Include them manually with the exception of some files in crypto:
> 
> unrelated question: Is there a reason they need to use the flag at all? The files are never compiled and linked when the flag isn't set, so making them appear "empty" when the flag isn't set is not needed?

Yeah, I agree. If there's some historical reason for this, I've forgotten it. It makes sense to me to make it either optionally compiled or optionally opted in (or out) via a define, but I don't see the need for both.

I do seem to remember some tools (`ar` or `ld`) that didn't like dealing with an empty file. So maybe it started as being unconditionally compiled with an ifdef to control, and became additionally guarded at the buildsystem level on top to account for those tools?

theuni commented on src/addrman.cpp:
```diff
@@ -3,6 +3,10 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#if defined(HAVE_CONFIG_H)
```
There are several headers (serialization comes to mind) that are very handy to use just by copying them out of our tree and using them as-is without a buildsystem. I do this quite often myself. I'd hate to give it up, but it's also become a goal of mine to remove the need for any autoconf defines for low-level files like that. Not a hard goal, just a nice-to-have. And with c++20 we're nearly there. See #29263 for a PR that does a good bit in this regard.

It would also be a nice property for libbitcoinkernel though :)

TheCharlatan commented:
Made this into a scripted-diff here: https://github.com/TheCharlatan/bitcoin/commit/e4186f771294a8b112bac759f92ea5e4bef8f168

There are no differences with the patch in this PR, besides a single whitespace issue. Feel free to pick the commit and make this into a scripted-diff as well, though I'm not sure if it is worth it, since the script is a bit dense.

epiccurious commented:
Concept ACK 70171e012a7cd6723aa2b07a3fb3d93920ec2010.

theuni commented:
> There doesn't seem to be any sane way to make this a scripted diff

@TheCharlatan saw this and said "hold my beer". Amazing!

theuni commented:
Replaced my commit with @TheCharlatan's artwork above.

It doesn't work in the c-i environment because `bitcoin-config.h.in` is missing. I added an `./autogen.sh` to the script but that doesn't help because that env doesn't have autoconf.

Next I tried prepending with a bare `autoheader`, which should be all that's actually needed, but that's missing too.

Sadly, I'm not sure it's worth continuing with this. Even if it does work, the verify script would have to modify the user's tree which isn't very nice :(

sedited committed:
scripted-diff: Fix bitcoin_config_h includes

-BEGIN VERIFY SCRIPT-

regex_string='^(?!//).*(AC_APPLE_UNIVERSAL_BUILD|BOOST_PROCESS_USE_STD_FS|CHAR_EQUALS_INT8|CLIENT_VERSION_BUILD|CLIENT_VERSION_IS_RELEASE|CLIENT_VERSION_MAJOR|CLIENT_VERSION_MINOR|COPYRIGHT_HOLDERS|COPYRIGHT_HOLDERS_FINAL|COPYRIGHT_HOLDERS_SUBSTITUTION|COPYRIGHT_YEAR|ENABLE_ARM_SHANI|ENABLE_AVX2|ENABLE_EXTERNAL_SIGNER|ENABLE_SSE41|ENABLE_TRACING|ENABLE_WALLET|ENABLE_X86_SHANI|ENABLE_ZMQ|HAVE_BOOST|HAVE_BUILTIN_CLZL|HAVE_BUILTIN_CLZLL|HAVE_BYTESWAP_H|HAVE_CLMUL|HAVE_CONSENSUS_LIB|HAVE_CXX20|HAVE_DECL_BE16TOH|HAVE_DECL_BE32TOH|HAVE_DECL_BE64TOH|HAVE_DECL_BSWAP_16|HAVE_DECL_BSWAP_32|HAVE_DECL_BSWAP_64|HAVE_DECL_FORK|HAVE_DECL_FREEIFADDRS|HAVE_DECL_GETIFADDRS|HAVE_DECL_HTOBE16|HAVE_DECL_HTOBE32|HAVE_DECL_HTOBE64|HAVE_DECL_HTOLE16|HAVE_DECL_HTOLE32|HAVE_DECL_HTOLE64|HAVE_DECL_LE16TOH|HAVE_DECL_LE32TOH|HAVE_DECL_LE64TOH|HAVE_DECL_PIPE2|HAVE_DECL_SETSID|HAVE_DECL_STRERROR_R|HAVE_DEFAULT_VISIBILITY_ATTRIBUTE|HAVE_DLFCN_H|HAVE_DLLEXPORT_ATTRIBUTE|HAVE_ENDIAN_H|HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR|HAVE_FDATASYNC|HAVE_GETENTROPY_RAND|HAVE_GETRANDOM|HAVE_GMTIME_R|HAVE_INTTYPES_H|HAVE_LIBADVAPI32|HAVE_LIBCOMCTL32|HAVE_LIBCOMDLG32|HAVE_LIBGDI32|HAVE_LIBIPHLPAPI|HAVE_LIBKERNEL32|HAVE_LIBOLE32|HAVE_LIBOLEAUT32|HAVE_LIBSHELL32|HAVE_LIBSHLWAPI|HAVE_LIBUSER32|HAVE_LIBUUID|HAVE_LIBWINMM|HAVE_LIBWS2_32|HAVE_MALLOC_INFO|HAVE_MALLOPT_ARENA_MAX|HAVE_MINIUPNPC_MINIUPNPC_H|HAVE_MINIUPNPC_UPNPCOMMANDS_H|HAVE_MINIUPNPC_UPNPERRORS_H|HAVE_NATPMP_H|HAVE_O_CLOEXEC|HAVE_POSIX_FALLOCATE|HAVE_PTHREAD|HAVE_PTHREAD_PRIO_INHERIT|HAVE_STDINT_H|HAVE_STDIO_H|HAVE_STDLIB_H|HAVE_STRERROR_R|HAVE_STRINGS_H|HAVE_STRING_H|HAVE_STRONG_GETAUXVAL|HAVE_SYSCTL|HAVE_SYSCTL_ARND|HAVE_SYSTEM|HAVE_SYS_ENDIAN_H|HAVE_SYS_PRCTL_H|HAVE_SYS_RESOURCES_H|HAVE_SYS_SELECT_H|HAVE_SYS_STAT_H|HAVE_SYS_SYSCTL_H|HAVE_SYS_TYPES_H|HAVE_SYS_VMMETER_H|HAVE_THREAD_LOCAL|HAVE_TIMINGSAFE_BCMP|HAVE_UNISTD_H|HAVE_VM_VM_PARAM_H|LT_OBJDIR|PACKAGE_BUGREPORT|PACKAGE_NAME|PACKAGE_STRING|PACKAGE_TARNAME|PACKAGE_URL|PACKAGE_VERSION|PTHREAD_CREATE_JOINABLE|QT_QPA_PLATFORM_ANDROID|QT_QPA_PLATFORM_COCOA|QT_QPA_PLATFORM_MINIMAL|QT_QPA_PLATFORM_WINDOWS|QT_QPA_PLATFORM_XCB|QT_STATICPLUGIN|STDC_HEADERS|STRERROR_R_CHAR_P|USE_ASM|USE_BDB|USE_DBUS|USE_NATPMP|USE_QRCODE|USE_SQLITE|USE_UPNP|_FILE_OFFSET_BITS|_LARGE_FILES)'

exclusion_files=":(exclude)src/minisketch :(exclude)src/crc32c :(exclude)src/secp256k1 :(exclude)src/crypto/sha256_arm_shani.cpp :(exclude)src/crypto/sha256_avx2.cpp :(exclude)src/crypto/sha256_sse41.cpp :(exclude)src/crypto/sha256_x86_shani.cpp"

git grep --perl-regexp --files-with-matches "$regex_string" -- '*.cpp' $exclusion_files | xargs git grep -L "bitcoin-config.h" | while read -r file; do line_number=$(awk -v my_file="$file" '/\/\/ file COPYING or https?:\/\/www.opensource.org\/licenses\/mit-license.php\./ {line = NR} /^\/\// && NR == line + 1 {while(getline && /^\/\//) line = NR} END {print line+1}' "$file"); sed -i "${line_number}i\\\\n\#if defined(HAVE_CONFIG_H)\\n#include <config/bitcoin-config.h>\\n\#endif" "$file"; done;

git grep --perl-regexp --files-with-matches "$regex_string" -- '*.h' $exclusion_files | xargs git grep -L "bitcoin-config.h" | while read -r file; do sed -i "/#define.*_H/a \\\\n\#if defined(HAVE_CONFIG_H)\\n#include <config/bitcoin-config.h>\\n\#endif" "$file"; done;

for file in $(git grep --files-with-matches 'bitcoin-config.h' -- '*.cpp' '*.h' $exclusion_files); do if ! grep -q --perl-regexp "$regex_string" $file; then sed -i '/HAVE_CONFIG_H/{N;N;N;d;}' $file; fi; done;

-END VERIFY SCRIPT-

The first command creates a regular expression for matching all bitcoin-config.h symbols in the following form: ^(?!//).*(AC_APPLE_UNIVERSAL_BUILD|BOOST_PROCESS_USE_STD_FS|...|_LARGE_FILES). It was generated with:
./autogen.sh && printf '^(?!//).*(%s)' $(awk '/^#undef/ {print $2}' src/config/bitcoin-config.h.in | paste -sd "|" -)

The second command holds a list of files and directories that should not be processed. These include subtree directories as well as some crypto files that already get their symbols through the makefile.

The third command checks for missing bitcoin-config headers in .cpp files and adds the header if it is missing.

The fourth command checks for missing bitcoin-config headers in .h files and adds the header if it is missing.

The fifth command checks for unneeded bitcoin-config headers in sources files and removes the header if it is unneeded.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/29404#issuecomment-1939796003

Could just replace the first command then. Not as nice, but for the purpose of this PR, reviewers can run the mentioned command manually: https://github.com/TheCharlatan/bitcoin/commit/b78e3ce41168600a1e73cb079b7eba07893753c4 (also applies some fixes, because the verify script uses `sh` and not `bash` :P).

theuni commented:
@TheCharlatan Sure, agreed it's better than nothing. Not quite self-contained, but given that a few of us have independently arrived at what the correct diff should be, it's easy to tell if it's correct.

Will take that commit.

hebasto reviewed (commented):
Concept ACK.

https://github.com/bitcoin/bitcoin/pull/29404#issuecomment-1933987554:
> So I guess we could add a little ... lint job for this instead?

Agree. Without a linter, the code might be drifted back in its current state.

theuni commented:
>     * The `timingsafe_bcmp` impl in the code was previously always picked. Now it may or may not be picked.

At least for now, this one should be coming from `crypto/common.h`. That wouldn't have been the case after #29263 though.

theuni commented:
PR title an description updated.

sedited reviewed (approved):
ACK 9d1dbbd4ceb8c04340927f5127195dc306adf3fc

hebasto reviewed (approved):
ACK 9d1dbbd4ceb8c04340927f5127195dc306adf3fc, I have reviewed the code and it looks OK.

fanquake commented:
Note there are at least two defines missing from the scriptied diff (`_TIME_BITS` & `__MINGW_USE_VC2005_COMPAT`), if using Autoconf 2.72 , however that should not make a difference here. Looks like `WORDS_BIGENDIAN` is also missing, but that should als be ok.

fanquake reviewed (approved):
ACK 9d1dbbd4ceb8c04340927f5127195dc306adf3fc
